src/core/libraries/usbd/usb_backend.h

```diff
@@ -0,0 +1,445 @@
+    bool s_has_removed_driver = false;
+};
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+#if defined(_WIN32)
+typedef CRITICAL_SECTION usbi_mutex_t;
+#else
+typedef pthread_mutex_t usbi_mutex_t;
+#endif
+
+struct UsbDeviceHandle {
+    usbi_mutex_t lock;
+    unsigned long claimed_interfaces;
+    struct list_head list;
+    struct libusb_device* dev;
+    int auto_detach_kernel_driver;
+};
+
+struct UsbDevice {
+    volatile long refcnt;
+
+    struct libusb_context* ctx;
+    struct libusb_device* parent_dev;
+
+    uint8_t bus_number;
+    uint8_t port_number;
+    uint8_t device_address;
+    enum libusb_speed speed;
+
+    struct list_head list;
+    unsigned long session_data;
+
+    struct libusb_device_descriptor device_descriptor;
+    volatile long attached;
+};
+
+class UsbEmulatedBackend : public UsbRealBackend {
+public:
+    s64 GetDeviceList(libusb_device*** list) override {
+        auto** fake = static_cast<libusb_device**>(calloc(2, sizeof(libusb_device*)));
+        fake[0] = GetDevice(nullptr);
+        fake[1] = nullptr;
+        *list = fake;
+
+        return 1;
+    }
+    void FreeDeviceList(libusb_device** list, s32 unref_devices) override {
+        libusb_free_device_list(list, 0); // to avoid unref fake device
+    }
+
+    s32 GetConfiguration(libusb_device_handle* dev, s32* config) override {
+        return LIBUSB_SUCCESS;
+    }
+    s32 GetActiveConfigDescriptor(libusb_device* dev, libusb_config_descriptor** config) override {
+        const auto endpoint_descs = FillEndpointDescriptorPair();
+        const auto interface_desc = FillInterfaceDescriptor(endpoint_descs);
+
+        const auto interface = new libusb_interface();
+        interface->altsetting = interface_desc;
+        interface->num_altsetting = 0;
+
+        const auto new_config = FillConfigDescriptor(interface);
+
+        ASSERT(endpoint_descs && interface_desc && new_config);
+        *config = new_config;
+        return LIBUSB_SUCCESS;
+    }
+    s32 GetConfigDescriptor(libusb_device* dev, u8 config_index,
+                            libusb_config_descriptor** config) override {
+        if (config_index >= 1) {
+            return LIBUSB_ERROR_NOT_FOUND;
+        }
+        return GetActiveConfigDescriptor(dev, config);
+    }
+
+    s32 GetMaxPacketSize(libusb_device* dev, u8 endpoint) override {
+        libusb_device_descriptor* desc = nullptr;
+
+        int r = GetDeviceDescriptor(dev, desc);
+        if (r < LIBUSB_SUCCESS) {
+            return LIBUSB_ERROR_OTHER;
+        }
+        return desc->bMaxPacketSize0;
+    }
+
+    s32 OpenDevice(libusb_device* dev, libusb_device_handle** dev_handle) override {
+        auto* _dev_handle = static_cast<UsbDeviceHandle*>(calloc(1, sizeof(libusb_device_handle*)));
+        if (!_dev_handle) {
+            return LIBUSB_ERROR_NO_MEM;
+        }
+        _dev_handle->dev = dev;
+        *dev_handle = reinterpret_cast<libusb_device_handle*>(_dev_handle);
```

Comment:
this cant be casted, libusb_device_handle and UsbDeviceHandle are not the same and it will segfault